client/asset/btc/btc.go

```diff
@@ -1447,31 +1595,118 @@ func (btc *ExchangeWallet) Confirmations(id dex.Bytes) (uint32, error) {
 // run pings for new blocks and runs the tipChange callback function when the
 // block changes.
 func (btc *ExchangeWallet) run(ctx context.Context) {
-	h, err := btc.node.GetBestBlockHash()
-	if err != nil {
-		btc.tipChange(fmt.Errorf("error initializing best block for %s: %v", btc.symbol, err))
-	}
-	tipHash := *h
 	ticker := time.NewTicker(blockTicker)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-ticker.C:
-			h, err := btc.node.GetBestBlockHash()
-			if err != nil {
-				btc.tipChange(fmt.Errorf("failed to get best block hash from %s node", btc.symbol))
-				continue
-			}
-			if *h != tipHash {
-				tipHash = *h
-				btc.tipChange(nil)
-			}
+			btc.checkForNewBlocks()
 		case <-ctx.Done():
 			return
 		}
 	}
 }
 
+// checkForNewBlocks checks for new blocks. When a tip change is detected, the
+// tipChange callback function is invoked and a goroutine is started to check
+// if any contracts in the findRedemptionQueue are redeemed in the new blocks.
+func (btc *ExchangeWallet) checkForNewBlocks() {
+	newTipHash, err := btc.node.GetBestBlockHash()
+	if err != nil {
+		btc.tipChange(fmt.Errorf("failed to get best block hash from %s node", btc.symbol))
+		return
+	}
+
+	// This method is called frequently. Don't hold write lock
+	// unless tip has changed.
+	btc.tipMtx.RLock()
+	sameTip := btc.currentTip.hash == newTipHash.String()
+	btc.tipMtx.RUnlock()
+	if sameTip {
+		return // tip has not changed
+	}
+
+	newTip, err := btc.blockFromHash(newTipHash.String())
+	if err != nil {
+		btc.tipChange(fmt.Errorf("error setting new tip: %v", err))
+		return
+	}
+
+	btc.tipMtx.Lock()
+	prevTip := btc.currentTip
+	btc.currentTip = newTip
+	btc.tipMtx.Unlock()
```

Comment:
> Tried the harness test, but ended up with
> 
> ```
> --- FAIL: TestTrading (24.85s)
>     --- FAIL: TestTrading/success (15.51s)
>         trade_simnet_test.go:205: btc balance change not in expected range -1.00180000 - -0.00180000, got 0.04816178
> ```
> 
> Confirmed on fresh harnesses. Is this related to #648?

Haven't narrowed it down but likely. The primary motivation for forcing coin unlocking for spent outputs in #648 was because of #453 and #471. The issue should be fixed now in dcrwallet anyways.

Reply:
`checkForNewBlocks` is called every second, not sure how I feel about locking and unlocking the tipMtx every second when most of the times `checkForNewBlocks` is called, we'd only be reading the current tip value to confirm there's a change.

Also, when a new block is noticed, findRedemptionMtx is locked for reads and I've ran into a situation where this caused a deadlock with btc.FindRedemption.

btc.FindRedemption locks the findRedemptionMtx when a new find redemption request is submitted. Later in that method, while the findRedemptionMtx is locked, btc.FindRedemption tries to acquire a read lock on the tipMtx.

If `checkForNewBlocks` notices a tip change at the same time `btc.FindRedemption` is called, `checkForNewBlocks` will wait for `btc.FindRedemption` to complete running in order to release the write lock on findRedemptionMtx, while `btc.FindRedemption` will not return because it is unable to acquire a read lock on tipMtx. Locking and unlocking the tipMtx only when needed and for the shortest amount possible allows `btc.FindRedemption` to read the newly set value, then unlock the findRedemptionMtx and then `checkForNewBlocks` can acquire the read lock on findRedemptionMtx.

That's a bit of a eye turner and can probably be simplified. Works for now but could come back to it after cleaning up other stuff.

Reply:
Got this sorted out, will push the commit soon. Still unsure about locking the tipMtx for writes every second when minutes would most likely pass before the write lock will be necessary; so the write lock and defered unlock are moved to after we're sure of tip change.